File: compiler/codegen/functions/SvFilterFunction.cpp
#include "SvFilterFunction.h"

#define _USE_MATH_DEFINES

#include "../MaximContext.h"
#include "../ComposableModuleClassMethod.h"
#include "../Num.h"
#include "../Tuple.h"

using namespace MaximCodegen;

SvFilterFunction::SvFilterFunction(MaximContext *ctx, llvm::Module *module)
    : Function(ctx, module, "svFilter",
               ctx->getTupleType({ctx->numType(), ctx->numType(), ctx->numType(), ctx->numType()}),
               {Parameter(ctx->numType(), false, false),
                Parameter(ctx->numType(), false, false),
                Parameter(ctx->numType(), false, false)},
               nullptr) {

}

std::unique_ptr<SvFilterFunction> SvFilterFunction::create(MaximContext *ctx, llvm::Module *module) {
    return std::make_unique<SvFilterFunction>(ctx, module);
}

std::unique_ptr<Value> SvFilterFunction::generate(ComposableModuleClassMethod *method,
                                                  const std::vector<std::unique_ptr<Value>> &params,
                                                  std::unique_ptr<VarArg> vararg) {
    auto sinFunc = llvm::Intrinsic::getDeclaration(method->moduleClass()->module(), llvm::Intrinsic::ID::sin,
                                                   {ctx()->numType()->vecType()});
    auto minFunc = llvm::Intrinsic::getDeclaration(method->moduleClass()->module(), llvm::Intrinsic::ID::minnum,
                                                   {ctx()->numType()->vecType()});
    auto powFunc = llvm::Intrinsic::getDeclaration(method->moduleClass()->module(), llvm::Intrinsic::ID::pow,
                                                   {ctx()->numType()->vecType()});

    auto inputNum = dynamic_cast<Num *>(params[0].get());
    auto freqNum = dynamic_cast<Num *>(params[1].get());
    auto qNum = dynamic_cast<Num *>(params[2].get());
    assert(inputNum && freqNum && qNum);

    auto &b = method->builder();
    auto notchPtr = method->getEntryPointer(method->moduleClass()->addEntry(ctx()->floatVecTy()), "notch.ptr");
    auto lowPtr = method->getEntryPointer(method->moduleClass()->addEntry(ctx()->floatVecTy()), "low.ptr");
    auto highPtr = method->getEntryPointer(method->moduleClass()->addEntry(ctx()->floatVecTy()), "high.ptr");
    auto bandPtr = method->getEntryPointer(method->moduleClass()->addEntry(ctx()->floatVecTy()), "band.ptr");

    auto inputVec = inputNum->vec(b);

    auto fParam = b.CreateFMul(
        freqNum->vec(b),
        b.CreateFDiv(ctx()->constFloatVec(M_PI), ctx()->constFloatVec(ctx()->sampleRate)),
        "fparam"
    );
    auto fOver2 = CreateCall(b, sinFunc, {fParam}, "fover2");
    auto fVal = b.CreateFMul(ctx()->constFloatVec(2), fOver2, "f");

    auto qVec = qNum->vec(b);
    auto qv = b.CreateFDiv(ctx()->constFloatVec(1), qVec);

    // calculate dampening factor
    auto twoQ = b.CreateFMul(qv, ctx()->constFloatVec(2), "twoq");
    auto twoQRecip = b.CreateFDiv(ctx()->constFloatVec(1), twoQ, "twoqrecip");
    auto dampPowBase = b.CreateFSub(ctx()->constFloatVec(1), twoQRecip, "damppowbase");
    auto dampPow = CreateCall(b, powFunc, {dampPowBase, ctx()->constFloatVec(0.25)}, "damppow");
    auto inverseDampPow = b.CreateFSub(ctx()->constFloatVec(1), dampPow, "inversedamppow");
    auto dampVal = b.CreateFMul(inverseDampPow, ctx()->constFloatVec(2), "dampval");

    auto maxDampLeft = b.CreateFDiv(ctx()->constFloatVec(2), fVal, "maxdamp.left");
    auto maxDampRight = b.CreateFMul(ctx()->constFloatVec(0.5), fVal, "maxdamp.right");
    auto maxDamp = b.CreateFSub(maxDampLeft, maxDampRight, "maxdamp");
    maxDamp = CreateCall(b, minFunc, {ctx()->constFloatVec(2), maxDamp}, "maxdamp");

    auto damp = CreateCall(b, minFunc, {dampVal, maxDamp}, "damp");

    auto loopIndexPtr = method->allocaBuilder().CreateAlloca(llvm::Type::getInt8Ty(ctx()->llvm()), nullptr,
                                                             "index.ptr");
    b.CreateStore(ctx()->constInt(8, 0, false), loopIndexPtr);
    auto loopCount = ctx()->constInt(8, 2, false);

    auto func = method->get(method->moduleClass()->module());
    auto loopCheckBlock = llvm::BasicBlock::Create(ctx()->llvm(), "loopcheck", func);
    auto loopBodyBlock = llvm::BasicBlock::Create(ctx()->llvm(), "loopbody", func);
    auto loopEndBlock = llvm::BasicBlock::Create(ctx()->llvm(), "loopend", func);

    b.CreateBr(loopCheckBlock);
    b.SetInsertPoint(loopCheckBlock);

    auto currentIndex = b.CreateLoad(loopIndexPtr, "currentindex");
    auto indexCond = b.CreateICmpULT(currentIndex, loopCount, "indexcond");
    b.CreateCondBr(indexCond, loopBodyBlock, loopEndBlock);

    b.SetInsertPoint(loopBodyBlock);

    b.CreateStore(
        b.CreateFSub(inputVec, b.CreateFMul(damp, b.CreateLoad(bandPtr)), "newnotch"),
        notchPtr
    );
    b.CreateStore(
        b.CreateFAdd(b.CreateLoad(lowPtr), b.CreateFMul(fVal, b.CreateLoad(bandPtr)), "newlow"),
        lowPtr
    );
    b.CreateStore(
        b.CreateFSub(b.CreateLoad(notchPtr), b.CreateLoad(lowPtr), "newhigh"),
        highPtr
    );
    b.CreateStore(
        b.CreateFAdd(b.CreateFMul(fVal, b.CreateLoad(highPtr)), b.CreateLoad(bandPtr), "newband"),
        bandPtr
    );

    auto nextIndex = b.CreateAdd(currentIndex, ctx()->constInt(8, 1, false), "indexptr");
    b.CreateStore(nextIndex, loopIndexPtr);

    b.CreateBr(loopCheckBlock);
    b.SetInsertPoint(loopEndBlock);

    auto inputForm = inputNum->form(b);
    auto inputActive = inputNum->active(b);

    Tuple::Storage tupleVecs;
    auto resultHigh = Num::create(ctx(), method->allocaBuilder());
    resultHigh->setVec(b, b.CreateLoad(highPtr, "highvec"));
    resultHigh->setForm(b, inputForm);
    resultHigh->setActive(b, inputActive);
    tupleVecs.push_back(std::move(resultHigh));

    auto resultLow = Num::create(ctx(), method->allocaBuilder());
    resultLow->setVec(b, b.CreateLoad(lowPtr, "lowvec"));
    resultLow->setForm(b, inputForm);
    resultLow->setActive(b, inputActive);
    tupleVecs.push_back(std::move(resultLow));

    auto resultBand = Num::create(ctx(), method->allocaBuilder());
    resultBand->setVec(b, b.CreateLoad(bandPtr, "bandvec"));
    resultBand->setForm(b, inputForm);
    resultBand->setActive(b, inputActive);
    tupleVecs.push_back(std::move(resultBand));

    auto resultNotch = Num::create(ctx(), method->allocaBuilder());
    resultNotch->setVec(b, b.CreateLoad(notchPtr, "notchvec"));
    resultNotch->setForm(b, inputForm);
    resultNotch->setActive(b, inputActive);
    tupleVecs.push_back(std::move(resultNotch));

    auto undefPos = SourcePos(-1, -1);
    return Tuple::create(
        ctx(),
        std::move(tupleVecs),
        b, method->allocaBuilder(), undefPos, undefPos
    );
}

File: compiler/src/codegen/functions/sv_filter_function.rs
use super::{Function, FunctionContext, VarArgs};
use codegen::values::{NumValue, TupleValue};
use codegen::{globals, intrinsics, util};
use inkwell::context::Context;
use inkwell::types::StructType;
use inkwell::values::PointerValue;
use inkwell::IntPredicate;
use mir::block;
use std::f32::consts;

pub struct SvFilterFunction {}
impl Function for SvFilterFunction {
    fn function_type() -> block::Function {
        block::Function::SvFilter
    }

    fn data_type(context: &Context) -> StructType {
        let float_vec = context.f32_type().vec_type(2);
        context.struct_type(
            &[
                &float_vec, // notch
                &float_vec, // low
                &float_vec, // high
                &float_vec, // band
            ],
            false,
        )
    }

    fn gen_call(
        func: &mut FunctionContext,
        args: &[PointerValue],
        _varargs: Option<VarArgs>,
        result: PointerValue,
    ) {
        let sin_intrinsic = intrinsics::sin_v2f32(func.ctx.module);
        let min_intrinsic = intrinsics::minnum_v2f32(func.ctx.module);
        let pow_intrinsic = intrinsics::pow_v2f32(func.ctx.module);

        let notch_ptr = unsafe { func.ctx.b.build_struct_gep(&func.data_ptr, 0, "notch.ptr") };
        let low_ptr = unsafe { func.ctx.b.build_struct_gep(&func.data_ptr, 1, "low.ptr") };
        let high_ptr = unsafe { func.ctx.b.build_struct_gep(&func.data_ptr, 2, "high.ptr") };
        let band_ptr = unsafe { func.ctx.b.build_struct_gep(&func.data_ptr, 3, "band.ptr") };

        let input_num = NumValue::new(args[0]);
        let freq_num = NumValue::new(args[1]);
        let q_num = NumValue::new(args[2]);
        let result_tuple = TupleValue::new(result);

        let input_vec = input_num.get_vec(func.ctx.b);
        let freq_vec = freq_num.get_vec(func.ctx.b);
        let f_val = func.ctx.b.build_float_mul(
            func.ctx
                .b
                .build_call(
                    &sin_intrinsic,
                    &[&func.ctx.b.build_float_mul(
                        freq_vec,
                        func.ctx.b.build_float_div(
                            util::get_vec_spread(func.ctx.context, consts::PI),
                            func.ctx
                                .b
                                .build_load(
                                    &globals::get_sample_rate(func.ctx.module).as_pointer_value(),
                                    "samplerate",
                                ).into_vector_value(),
                            "",
                        ),
                        "fparam",
                    )],
                    "",
                    false,
                ).left()
                .unwrap()
                .into_vector_value(),
            util::get_vec_spread(func.ctx.context, 2.),
            "f",
        );

        let q_vec = q_num.get_vec(func.ctx.b);
        let q_v =
            func.ctx
                .b
                .build_float_div(util::get_vec_spread(func.ctx.context, 1.), q_vec, "qv");

        // calculate dampening factor
        let damp_val = func.ctx.b.build_float_mul(
            func.ctx.b.build_float_sub(
                util::get_vec_spread(func.ctx.context, 1.),
                func.ctx
                    .b
                    .build_call(
                        &pow_intrinsic,
                        &[
                            &func.ctx.b.build_float_sub(
                                util::get_vec_spread(func.ctx.context, 1.),
                                func.ctx.b.build_float_div(
                                    util::get_vec_spread(func.ctx.context, 1.),
                                    func.ctx.b.build_float_mul(
                                        q_v,
                                        util::get_vec_spread(func.ctx.context, 2.),
                                        "twoq",
                                    ),
                                    "twoqrecip",
                                ),
                                "damppowbase",
                            ),
                            &util::get_vec_spread(func.ctx.context, 0.25),
                        ],
                        "damppow",
                        false,
                    ).left()
                    .unwrap()
                    .into_vector_value(),
                "inversedamppow",
            ),
            util::get_vec_spread(func.ctx.context, 2.),
            "dampval",
        );

        let max_damp = func
            .ctx
            .b
            .build_call(
                &min_intrinsic,
                &[
                    &util::get_vec_spread(func.ctx.context, 2.),
                    &func.ctx.b.build_float_sub(
                        func.ctx.b.build_float_div(
                            util::get_vec_spread(func.ctx.context, 2.),
                            f_val,
                            "maxdamp.left",
                        ),
                        func.ctx.b.build_float_mul(
                            util::get_vec_spread(func.ctx.context, 0.5),
                            f_val,
                            "maxdamp.right",
                        ),
                        "maxdamp",
                    ),
                ],
                "maxdamp",
                false,
            ).left()
            .unwrap()
            .into_vector_value();
        let damp = func
            .ctx
            .b
            .build_call(&min_intrinsic, &[&damp_val, &max_damp], "damp", false)
            .left()
            .unwrap()
            .into_vector_value();

        let loop_index_ptr = func
            .ctx
            .allocb
            .build_alloca(&func.ctx.context.i8_type(), "index.ptr");
        func.ctx.b.build_store(
            &loop_index_ptr,
            &func.ctx.context.i8_type().const_int(0, false),
        );

        let loop_check_block = func
            .ctx
            .context
            .append_basic_block(&func.ctx.func, "loopcheck");
        let loop_body_block = func
            .ctx
            .context
            .append_basic_block(&func.ctx.func, "loopbody");
        let loop_end_block = func
            .ctx
            .context
            .append_basic_block(&func.ctx.func, "loopend");
        func.ctx.b.build_unconditional_branch(&loop_check_block);

        func.ctx.b.position_at_end(&loop_check_block);
        let current_index = func
            .ctx
            .b
            .build_load(&loop_index_ptr, "index")
            .into_int_value();
        let index_cond = func.ctx.b.build_int_compare(
            IntPredicate::ULT,
            current_index,
            func.ctx.context.i8_type().const_int(2, false),
            "indexcond",
        );
        func.ctx
            .b
            .build_conditional_branch(&index_cond, &loop_body_block, &loop_end_block);

        func.ctx.b.position_at_end(&loop_body_block);
        func.ctx.b.build_store(
            &notch_ptr,
            &func.ctx.b.build_float_sub(
                input_vec,
                func.ctx.b.build_float_mul(
                    damp,
                    func.ctx.b.build_load(&band_ptr, "band").into_vector_value(),
                    "",
                ),
                "newnotch",
            ),
        );
        func.ctx.b.build_store(
            &low_ptr,
            &func.ctx.b.build_float_add(
                func.ctx.b.build_load(&low_ptr, "low").into_vector_value(),
                func.ctx.b.build_float_mul(
                    f_val,
                    func.ctx.b.build_load(&band_ptr, "band").into_vector_value(),
                    "",
                ),
                "newlow",
            ),
        );
        func.ctx.b.build_store(
            &high_ptr,
            &func.ctx.b.build_float_sub(
                func.ctx
                    .b
                    .build_load(&notch_ptr, "notch")
                    .into_vector_value(),
                func.ctx.b.build_load(&low_ptr, "low").into_vector_value(),
                "newhigh",
            ),
        );
        func.ctx.b.build_store(
            &band_ptr,
            &func.ctx.b.build_float_add(
                func.ctx.b.build_float_mul(
                    f_val,
                    func.ctx.b.build_load(&high_ptr, "high").into_vector_value(),
                    "",
                ),
                func.ctx.b.build_load(&band_ptr, "band").into_vector_value(),
                "newband",
            ),
        );

        let next_index = func.ctx.b.build_int_add(
            current_index,
            func.ctx.context.i8_type().const_int(1, false),
            "nextindex",
        );
        func.ctx.b.build_store(&loop_index_ptr, &next_index);
        func.ctx.b.build_unconditional_branch(&loop_check_block);

        func.ctx.b.position_at_end(&loop_end_block);

        let input_form = input_num.get_form(func.ctx.b);

        let high_result_num = NumValue::new(result_tuple.get_item_ptr(func.ctx.b, 0));
        let high_result_vec = func.ctx.b.build_load(&high_ptr, "high").into_vector_value();
        high_result_num.set_vec(func.ctx.b, &high_result_vec);
        high_result_num.set_form(func.ctx.b, &input_form);

        let low_result_num = NumValue::new(result_tuple.get_item_ptr(func.ctx.b, 1));
        let low_result_vec = func.ctx.b.build_load(&low_ptr, "low").into_vector_value();
        low_result_num.set_vec(func.ctx.b, &low_result_vec);
        low_result_num.set_form(func.ctx.b, &input_form);

        let band_result_num = NumValue::new(result_tuple.get_item_ptr(func.ctx.b, 2));
        let band_result_vec = func.ctx.b.build_load(&band_ptr, "band").into_vector_value();
        band_result_num.set_vec(func.ctx.b, &band_result_vec);
        band_result_num.set_form(func.ctx.b, &input_form);

        let notch_result_num = NumValue::new(result_tuple.get_item_ptr(func.ctx.b, 3));
        let notch_result_vec = func
            .ctx
            .b
            .build_load(&notch_ptr, "notch")
            .into_vector_value();
        notch_result_num.set_vec(func.ctx.b, &notch_result_vec);
        notch_result_num.set_form(func.ctx.b, &input_form);
    }
}
